### ISSUE TITLE
After deleting, adding again and decompressing will report an error

### ISSUE THREAD

**Issue by YSLdragon** [srikanth-lingala/zip4j#236]
First, I decompress and delete some files from the zip file, and then modify them and add them to the zip file. In this way, there is no problem with the zip file, but this error will be reported when decompressing again

`Exception in thread "main" net.lingala.zip4j.exception.ZipException: File header and local file header mismatch
	at net.lingala.zip4j.tasks.AbstractExtractFileTask.verifyNextEntry(AbstractExtractFileTask.java:144)
	at net.lingala.zip4j.tasks.AbstractExtractFileTask.extractFile(AbstractExtractFileTask.java:52)
	at net.lingala.zip4j.tasks.ExtractFileTask.executeTask(ExtractFileTask.java:39)
	at net.lingala.zip4j.tasks.ExtractFileTask.executeTask(ExtractFileTask.java:21)
	at net.lingala.zip4j.tasks.AsyncZipTask.performTaskWithErrorHandling(AsyncZipTask.java:46)
	at net.lingala.zip4j.tasks.AsyncZipTask.execute(AsyncZipTask.java:40)
	at net.lingala.zip4j.ZipFile.extractFile(ZipFile.java:485)
	at net.lingala.zip4j.ZipFile.extractFile(ZipFile.java:451)
	at main.unzipf(main.java:71)
	at main.process(main.java:99)
	at main.main(main.java:25)`

**Comment by srikanth-lingala** [comment]
Based on your description, I have written a small program to test this. It worked fine for me. You can test it with the attached test.zip file, which is just the zip4j jar renamed to test.zip.

```
package net.lingala.zip4j;

import net.lingala.zip4j.model.FileHeader;
import net.lingala.zip4j.model.ZipParameters;
import org.assertj.core.api.Assertions;
import org.junit.Test;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Properties;

public class ZipTest {

    private static final String BASE_DIR = "/Users/srikanth/test/";
    private static final String OUTPUT_DIR = BASE_DIR + "output/";

    @Test
    public void testIssue236() throws IOException {
        ZipFile zipFile = new ZipFile(BASE_DIR + "test.zip");
        String fileToEdit = "META-INF/maven/net.lingala.zip4j/zip4j/pom.properties";
        String versionNumberToSet = "2.7.2";

        extractFile(zipFile, fileToEdit);
        removeFileFromZip(zipFile, fileToEdit);
        editFile(fileToEdit, versionNumberToSet);
        addFileToZip(zipFile, fileToEdit);
        extractFile(zipFile, fileToEdit);
        verifyEditedProperties(fileToEdit, versionNumberToSet);
    }

    private void extractFile(ZipFile zipFile, String fileToExtract) throws IOException {
        FileHeader fileHeader = zipFile.getFileHeader(fileToExtract);

        if (fileHeader == null) {
            throw new RuntimeException("Could not find a file by name " + fileToExtract + " in the zip file");
        }
        System.out.println("Extracting file " + fileToExtract);
        zipFile.extractFile(fileHeader, OUTPUT_DIR);
    }

    private void removeFileFromZip(ZipFile zipFile, String fileToRemove) throws IOException {
        System.out.println("Removing " + fileToRemove + " from zip file " + zipFile.getFile());
        zipFile.removeFile(fileToRemove);
    }

    private void editFile(String fileToEdit, String versionNumberToSet) throws IOException {
        Properties properties = readProperties(fileToEdit);
        properties.setProperty("version", versionNumberToSet);
        try (OutputStream outputStream = new FileOutputStream(new File(OUTPUT_DIR + fileToEdit))) {
            System.out.println("Saving edited properties file " + fileToEdit);
            properties.store(outputStream, "");
        }
    }

    private void addFileToZip(ZipFile zipFile, String fileToAdd) throws IOException {
        ZipParameters zipParameters = new ZipParameters();
        zipParameters.setFileNameInZip(fileToAdd);
        System.out.println("Adding edited file " + fileToAdd + " to zip " + zipFile.getFile());
        zipFile.addFile(BASE_DIR + "output/" + fileToAdd, zipParameters);
    }

    private void verifyEditedProperties(String fileToVerify, String versionNumberToSet) throws IOException {
        Properties properties = readProperties(fileToVerify);
        Assertions.assertThat(properties.getProperty("version")).isEqualTo(versionNumberToSet);
    }

    private Properties readProperties(String propertiesFile) throws IOException {
        Properties properties = new Properties();

        try (InputStream inputStream = new FileInputStream(new File(OUTPUT_DIR + propertiesFile))) {
            properties.load(inputStream);
        }

        return properties;
    }
}

```

[test.zip](https://github.com/srikanth-lingala/zip4j/files/5196715/test.zip)

**Comment by YSLdragon** [comment]
I tried again later. The reason for the error is that the original zip package does not have a header for the folder, so the first decompression will not make an error. However, after re adding the zip package, the header corresponding to a new folder will be generated. However, the folder cannot be decompressed as a single file, so an error occurs. When decompressing, add the judgment of the folder header to solve this problem. Thank you very much for your answer

**Comment by YSLdragon** [comment]
Now I encounter a new problem, or the previous process, decompress a folder from the compressed file and add it to the original compressed file. The main problem this time is about the running time of the program. If I do this directly on an APK package, the running time is very short. However, if the same APK package is decompressed and compressed (the content remains unchanged), the operation time of the appeal operation will be doubled, and even run for several minutes in some larger APK packages. I want to know what causes the time difference of the above operations. I have made a specific comparison. The main difference is that it takes longer to compress the folder into APK again.

**Comment by srikanth-lingala** [comment]
Correct me if I am wrong, but the way I understand your question is: why does it take much longer to unpack and pack the content of the zip file again when compared to just deleting and adding a single file. That is true, because when zip4j has to delete/replace a single file in the zip, zip4j copies the rest of the content as it is, and just adds/deletes the single file to be replaced. Whereas when unpacking and packing the complete zip, zip4j has to decompress all the files and compress all the files in the zip again. 

For example, if a file in a zip start at file location 1000th byte to 10000th byte and the total file size of zip is 20000 bytes, zip4j will copy from 1st byte to 999th byte "as it is", then processes byte 1000 to 10000, and then copies again 10001th byte to 20000 bytes "as it is". This will save a lot of time when compared to unpacking and packing when zip4j will have to process completely from 1st byte to 20000th byte.

Hope it makes sense.

**Comment by YSLdragon** [comment]
http://quality-cdn.shouji.sogou.com/wapdl/android/apk/SogouInput_android_v10.16_sweb.apk?frm=new_pcjs_index
In fact, there is such a special type of zip file that we call APK. I use the following handler (I mentioned above) to process it the first time, but it takes several times as long to process the result of the first processing again. The content of the file has not changed in the whole process, so I don't know why there is such a difference. Since I can't upload test files larger than 10m, I put a URL link to download an APK file.

`    package main;
    import net.lingala.zip4j.ZipFile;
    import net.lingala.zip4j.exception.ZipException;
    import net.lingala.zip4j.model.FileHeader;
    import org.apache.commons.io.FileUtils;
    import java.io.File;
    import java.io.IOException;
    import java.util.ArrayList;
    import java.util.List;

    public class test {
    public static String mainpath=System.getProperty("user.dir");

    public static void main(String[] args) throws IOException {
        String apkpath=args[0];
        process(apkpath);
    }

    public static List<String> unzipf(String apkpath, String unzipath) throws ZipException {
        ZipFile zipFile = new ZipFile(apkpath);
        List fileHeaderList = zipFile.getFileHeaders();
        List<String> rNames = new ArrayList<String>();
        for (int i = 0; i < fileHeaderList.size(); i++) {
            FileHeader fileHeader = (FileHeader)fileHeaderList.get(i);
            if(!fileHeader.getFileName().startsWith("res/")||fileHeader.isDirectory()) continue;
            System.out.println(fileHeader.getFileName());
            rNames.add(fileHeader.getFileName());
            zipFile.extractFile(fileHeader, unzipath);
        }
        return rNames;
    }

    public static List<String> getziph(String zipath) throws ZipException {
        ZipFile zipFile = new ZipFile(zipath);

        List headersList = zipFile.getFileHeaders();
        List<String> rNames = new ArrayList<String>();
        for(int i=0, len = headersList.size(); i<len; i++) {
            FileHeader subHeader = (FileHeader) headersList.get(i);
            if(!subHeader.getFileName().startsWith("lib")) continue;
            rNames.add(subHeader.getFileName());
        }
        for(String hname : rNames) {
            System.out.println(hname);
        }
        return rNames;
    }


    public static void process(String apkpath) throws ZipException {
        String workpath=mainpath+"\\tmpfile";
        String unzipath=workpath+"\\";
        File unzf=new File(unzipath+"res");
        if(unzf.exists()) FileUtils.deleteQuietly(unzf);
        long time1=System.currentTimeMillis();
        List<String> namelist =unzipf(apkpath,unzipath);
        long time2=System.currentTimeMillis();
        ZipFile zipFile = new ZipFile(apkpath);
        zipFile.removeFiles(namelist);
        long time3=System.currentTimeMillis();
       zipFile.addFolder(unzf);
        long time4=System.currentTimeMillis();
        System.out.println("unzip time:"+(time2-time1)+"ms");
        System.out.println("delete time:"+(time3-time2)+"ms");
        System.out.println("add time:"+(time4-time3)+"ms");
    }
    public static ArrayList<String> getFiles(String path) {
        ArrayList<String> files = new ArrayList<String>();
        File file = new File(path);
        File[] tempList = file.listFiles();
        for (int i = 0; i < tempList.length; i++) {
            if (tempList[i].isFile()) {
                files.add(tempList[i].toString());
                //System.out.println(tempList[i].toString());
            } else {
                //System.out.println("add folder "+tempList[i].getName());
                files.add(tempList[i].toString());
                getFile(path + File.separator + tempList[i].getName(), files);
            }
        }
        return files;
    }

    public static ArrayList<String> getFile(String path, ArrayList<String> files) {
        File file = new File(path);
        File[] tempList = file.listFiles();
        for (int i = 0; i < tempList.length; i++) {
            if (tempList[i].isFile()) {
                files.add(tempList[i].toString());
                //System.out.println(tempList[i].toString());
            }
        }
        return files;
    }
}
`

**Comment by srikanth-lingala** [comment]
I tried your code, and I will repeat the point I made in my previous post in this issue. The reason it takes longer to re-zip the files when compared to deleting the files is because there is more processing involved when you zip a file than when you delete content from a zip file. Deleting an entry in the zip file just removes the corresponding entry in the zip file without processing any other file in the zip. Whereas, adding a zip file does a lot more processing (Inflating, etc) of each entry. You have more than 7000 files that are added to the zip file, and adding so many files to the zip takes a bit longer than when you just remove files from the zip.
